### ISSUE TITLE
No CANCELED state is received  when canceling an upload

### ISSUE THREAD

**Issue by gilbertdigio** [aws-amplify/aws-sdk-android#676]
**Describe the bug**

When uploading multiple videos in parallel some of them go to WAITING status. If any of those videos is canceled, we stop receiving updates from TransferListener for that transfer. Neither the CANCELED state nor RESUMED_WAITING.

**To Reproduce**
1. Upload 10 or more videos in parallel.
2. Cancel 1 or more of those transfers while they are in WAITING state.
3. Note that stop receiving state updates from TransferListener.

**Which AWS service(s) are affected?**
Transfer Utility from android phone

**Expected behavior**
The transfer should be canceled even if it's in WAITING state.

**Environment(please complete the following information):**
 - SDK Version: 2.11.0

**Device Information (please complete the following information):**
 - Device: Samsung Galaxy S8
 - Android Version: Oreo

**Comment by kvasukib** [comment]
@gilbertdigio Sorry for the inconvenience caused. I wrote a test to see if I could reproduce this issue. Basically, I restricted the thread pool to 2 threads and started 10 uploads and canceled random uploads. I did see onStateChanged with CANCELED.

Here is the test:

```java
    @Test
    public void testCancelWhenTransfersAreInWaitingState() throws Exception {
        TransferUtilityOptions transferUtilityOptions = new TransferUtilityOptions();
        transferUtilityOptions.setTransferThreadPoolSize(2);

        TransferUtility transferUtilityWithReducedPoolSize = TransferUtility.builder()
                .context(appContext)
                .defaultBucket(BUCKET_NAME)
                .s3Client(sS3Client)
                .transferUtilityOptions(transferUtilityOptions)
                .build();

        List<TransferObserver> observers = new ArrayList<TransferObserver>();
        for (int i = 0; i < 10; i++) {
            final long size = 10 * MB;
            final File tempFile = temp.newFile();
            fillFile(tempFile, size);
            TransferObserver observer = transferUtility.upload("testCancelWhenTransfersAreInWaitingState-" + tempFile.getName(), tempFile);
            observer.setTransferListener(new TransferListener() {
                @Override
                public void onStateChanged(int i, TransferState transferState) {
                    Log.d(LOG_TAG, "onStateChanged: State = " + transferState);
                }

                @Override
                public void onProgressChanged(int i, long l, long l1) {
                    Log.d(LOG_TAG, "onProgressChanged: current = " + l + "; total = " + l1);
                }

                @Override
                public void onError(int i, Exception e) {
                    Log.e(LOG_TAG, "error = " + e);
                }
            });
            observers.add(observer);
        }

        for (TransferObserver observer : observers) {
            observer.refresh();
            Log.d(LOG_TAG, observer.toString());
        }

        Log.d(LOG_TAG, "Cancel status for 4: " + transferUtility.cancel(observers.get(4).getId()));
        Log.d(LOG_TAG, "Cancel status for 5: " + transferUtility.cancel(observers.get(5).getId()));
        Log.d(LOG_TAG, "Cancel status for 7: " + transferUtility.cancel(observers.get(7).getId()));

        for (TransferObserver observer : observers) {
            observer.refresh();
            Log.d(LOG_TAG, "State = " + observer.getState());
        }
    }
```


```

01-25 11:01:31.830 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=132, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit4544379855499831947.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=IN_PROGRESS, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit4544379855499831947.tmp'}
01-25 11:01:31.831 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=135, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit4285945899776150051.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=IN_PROGRESS, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit4285945899776150051.tmp'}
01-25 11:01:31.833 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=138, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit2110075397411085583.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=IN_PROGRESS, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit2110075397411085583.tmp'}
01-25 11:01:31.835 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=141, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit4609396763464476699.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit4609396763464476699.tmp'}
01-25 11:01:31.838 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=144, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit7033388820105867800.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit7033388820105867800.tmp'}
01-25 11:01:31.840 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=147, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit3381420071538616018.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit3381420071538616018.tmp'}
01-25 11:01:31.842 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=150, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit561709369687379482.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit561709369687379482.tmp'}
01-25 11:01:31.844 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=153, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit195865482393069906.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit195865482393069906.tmp'}
01-25 11:01:31.846 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=156, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit4866806571385698731.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit4866806571385698731.tmp'}
01-25 11:01:31.848 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: TransferObserver{id=159, bucket='tu-sample', key='testCancelWhenTransfersAreInWaitingState-junit541517686157748462.tmp', bytesTotal=10485760, bytesTransferred=0, transferState=WAITING, filePath='/data/user/0/com.amazonaws.s3transferutilitysample/cache/junit4887102580077574605/junit541517686157748462.tmp'}
```

```
01-25 11:01:29.898 6263-6263/? D/TransferUtilityInstrumentedTest: onStateChanged: State = IN_PROGRESS
01-25 11:01:30.117 6263-6263/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: onStateChanged: State = IN_PROGRESS
01-25 11:01:30.419 6263-6263/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: onStateChanged: State = IN_PROGRESS
01-25 11:01:31.859 6263-6263/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: onStateChanged: State = CANCELED
01-25 11:01:31.888 6263-6263/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: onStateChanged: State = CANCELED
01-25 11:01:31.910 6263-6263/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: onStateChanged: State = CANCELED
01-25 11:01:31.912 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = IN_PROGRESS
01-25 11:01:31.917 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = IN_PROGRESS
01-25 11:01:31.921 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = IN_PROGRESS
01-25 11:01:31.927 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = WAITING
01-25 11:01:31.944 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = CANCELED
01-25 11:01:31.945 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = CANCELED
01-25 11:01:31.948 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = WAITING
01-25 11:01:31.963 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = CANCELED
01-25 11:01:31.970 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = WAITING
01-25 11:01:31.973 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: State = WAITING
```

```
01-25 11:01:31.859 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: Cancel status for 4: true
01-25 11:01:31.889 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: Cancel status for 5: true
01-25 11:01:31.910 6263-6279/com.amazonaws.s3transferutilitysample D/TransferUtilityInstrumentedTest: Cancel status for 7: true
```

Can you check my test and let me know if I'm missing something in the attempt to reproduce it?

**Comment by gilbertdigio** [comment]
Thanks @kvasukib, 

I didn't refresh the observer for the WAITING state case.

 Fixed.

**Comment by kvasukib** [comment]
@gilbertdigio Thank you for confirming. Please feel free to open a new issue when you observer any issues in canceling a transfer while testing.
